GPU/Common/DrawEngineCommon.cpp

```diff
@@ -731,69 +597,121 @@ u32 DrawEngineCommon::ComputeMiniHash() {
+	return x;
+}
+
+uint32_t DrawEngineCommon::ComputeDrawcallsHash() const {
+	uint32_t dcid = 0;
+	for (int i = 0; i < numDrawVerts_; i++) {
+		u32 dhash = dcid;
+		dhash = __rotl(dhash ^ (u32)(uintptr_t)drawVerts_[i].verts, 13);
+		dhash = __rotl(dhash ^ (u32)drawInds_[i].vertexCount, 11);
+		dcid = lowbias32_r(dhash ^ (u32)drawInds_[i].prim);
+	}
+	for (int j = 0; j < numDrawInds_; j++) {
+		u32 dhash = dcid;
+		dhash = __rotl(dhash ^ (u32)(uintptr_t)drawInds_[j].inds, 19);
+		dcid = lowbias32_r(__rotl(dhash ^ (u32)drawInds_[j].indexType, 7));
+	}
+	return dcid;
+}
+
+int DrawEngineCommon::ComputeNumVertsToDecode() const {
+	int sum = 0;
+	for (int i = 0; i < numDrawVerts_; i++) {
+		sum += drawVerts_[i].indexUpperBound + 1 - drawVerts_[i].indexLowerBound;
+	}
+	return sum;
+}
+
 uint64_t DrawEngineCommon::ComputeHash() {
 	uint64_t fullhash = 0;
 	const int vertexSize = dec_->GetDecVtxFmt().stride;
-	const int indexSize = IndexSize(dec_->VertexType());
 
 	// TODO: Add some caps both for numDrawCalls_ and num verts to check?
 	// It is really very expensive to check all the vertex data so often.
-	for (int i = 0; i < numDrawCalls_; i++) {
-		const DeferredDrawCall &dc = drawCalls_[i];
-		if (!dc.inds) {
-			fullhash += XXH3_64bits((const char *)dc.verts, vertexSize * dc.vertexCount);
-		} else {
-			int indexLowerBound = dc.indexLowerBound, indexUpperBound = dc.indexUpperBound;
-			int j = i + 1;
-			int lastMatch = i;
-			while (j < numDrawCalls_) {
-				if (drawCalls_[j].verts != dc.verts)
-					break;
-				indexLowerBound = std::min(indexLowerBound, (int)dc.indexLowerBound);
-				indexUpperBound = std::max(indexUpperBound, (int)dc.indexUpperBound);
-				lastMatch = j;
-				j++;
-			}
-			// This could get seriously expensive with sparse indices. Need to combine hashing ranges the same way
-			// we do when drawing.
-			fullhash += XXH3_64bits((const char *)dc.verts + vertexSize * indexLowerBound,
-				vertexSize * (indexUpperBound - indexLowerBound));
+	for (int i = 0; i < numDrawVerts_; i++) {
+		const DeferredVerts &dv = drawVerts_[i];
+		int indexLowerBound = dv.indexLowerBound, indexUpperBound = dv.indexUpperBound;
+		fullhash += XXH3_64bits((const char *)dv.verts + vertexSize * indexLowerBound, vertexSize * (indexUpperBound - indexLowerBound));
+	}
+
+	for (int i = 0; i < numDrawInds_; i++) {
+		const DeferredInds &di = drawInds_[i];
+		if (di.indexType != 0) {
+			int indexSize = IndexSize(di.indexType << GE_VTYPE_IDX_SHIFT);
 			// Hm, we will miss some indices when combining above, but meh, it should be fine.
-			fullhash += XXH3_64bits((const char *)dc.inds, indexSize * dc.vertexCount);
-			i = lastMatch;
+			fullhash += XXH3_64bits((const char *)di.inds, indexSize * di.vertexCount);
 		}
 	}
 
-	fullhash += XXH3_64bits(&drawCalls_[0].uvScale, sizeof(drawCalls_[0].uvScale) * numDrawCalls_);
+	// this looks utterly broken??
+	// fullhash += XXH3_64bits(&drawCalls_[0].uvScale, sizeof(drawCalls_[0].uvScale) * numDrawCalls_);
```

Comment:
Looks like this was introduced all the way back in 4cf5999bc81b65386c75cca503e56f9831a2c5eb.  Probably it'd be better to just hash this within the loop now.  I suppose it still should be hashed...

-[Unknown]

Reply:
Hah, yeah my bad it seems. I'm gonna tweak the hashes a bit soon...

Reply:
Also the vertex cache benefit seems smaller than ever with the latest optimizations, so another option is to just rip it out...

Reply:
One of these years, we'll finally remove the vertex cache.

-[Unknown]